Enterprise Release Management changes

gherceg opened the pull request:
## Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/SS-148
https://dimagi-dev.atlassian.net/browse/SS-103
https://dimagi-dev.atlassian.net/browse/SS-147
https://dimagi-dev.atlassian.net/browse/SS-160
https://dimagi-dev.atlassian.net/browse/SS-187
https://dimagi-dev.atlassian.net/browse/SS-167
https://dimagi-dev.atlassian.net/browse/SS-173
https://dimagi-dev.atlassian.net/browse/SS-195
https://dimagi-dev.atlassian.net/browse/SS-128

This is the parent branch that encompasses the majority of work done for ERM, including all UI changes, and changes around permissions to ERM vs the feature flag version of linked projects. This PR does not contain any code that has not been reviewed prior, and this branch has gone through QA. 

Lessons learned: 
- improve naming for a base branch in the future (didn't expect this to become a base branch when first created)
- track under a Story in JIRA
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
`LINKED_DOMAINS` and `ERM_DEVELOPMENT`
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Wrote tests wherever appropriate along the way.
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi-dev.atlassian.net/browse/QA-3157
### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
This PR is large because it is the parent branch of many smaller PRs.
### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->
The PR linked [here](https://github.com/dimagi/commcare-hq/pull/30198) should be reverted prior to reverting this PR.
- [x] This PR can be reverted after deploy with no further considerations

gherceg committed:
add pagination to domain links table

gherceg committed:
rename Delete to Remove

gherceg committed:
fix total items bug

gherceg committed:
use modals correctly

gherceg committed:
ensure all text is marked with trans

gherceg committed:
address feedback

- change btn-primary to btn-danger
- domain_links doesn't need to be observable
- folow js naming scheme for some vars
- translate aria label

gherceg committed:
appease stickler

gherceg committed:
initial UI work

- add downstream button
- trigger modal

gherceg committed:
add method to retrieve available domains to link

gherceg committed:
address feedback correctly

gherceg committed:
setup method to handle adding new link on the frontend

gherceg committed:
fix select2 usage

gherceg committed:
remove from availableDomains

not sure why I didn't do that before

gherceg committed:
put button behind feature flag

gherceg committed:
add href field

gherceg committed:
Merge pull request #29826 from dimagi/gh/add-pagination

Add pagination to linked/downstream projects tab

gherceg committed:
I'm sorry stickler

gherceg committed:
add panel and spacer

gherceg committed:
add search box with filter logic

gherceg committed:
address feedback

- put text into a blocktrans
- use target="_blank" for help link
- rename test_db_accessors.py to test_dbaccessors.py
- remove newlyAddedDomainLinks since it is unnecessary

gherceg committed:
what would I do without you stickler

gherceg committed:
basic implementation to create domain link via rmi

gherceg committed:
update and sort availableDomains when removing and adding links

gherceg committed:
Merge branch 'gh/add-downstream-domain-btn' into gh/add-search-box

gherceg committed:
forgot a semicolon

gherceg committed:
nested div incorrectly

gherceg committed:
Merge pull request #29847 from dimagi/gh/add-downstream-domain-btn

Add downstream domain UI work

gherceg committed:
Merge pull request #29855 from dimagi/gh/add-search-box

Add search box to Linked Projects page

gherceg committed:
stickler indentation issue

gherceg committed:
update text to display which domain is being removed

gherceg committed:
put each tab into separate file

gherceg committed:
update text, add modal

gherceg committed:
add panel

gherceg committed:
add search and pagination; minor js refactor

gherceg committed:
grab upstreamLink in pullReleaseContentViewModel

gherceg committed:
trailing commas

gherceg committed:
address feedback and minor edits

- check success in js response
- translate 'Never'
- refactor link_context to clearer name
- display exception message to user
- minor stickler complaints

gherceg committed:
PullReleaseContentViewModel does not use data.can_update

gherceg committed:
use _.template to translate dynamic string

gherceg committed:
pass in timezone object

gherceg committed:
fix views.py import statements

gherceg committed:
Merge pull request #29881 from dimagi/gh/rmi-create-domain-link

[ERM UI] Create domain link RMI

gherceg committed:
update text

gherceg committed:
Merge pull request #29882 from dimagi/gh/remove-link-modal-v2

[ERM UI] Update text to display which domain is being removed

gherceg committed:
update text

gherceg committed:
Merge branch 'gh/improve-erm-ui' into gh/update-text-on-release-content

gherceg committed:
Merge branch 'gh/update-text-on-release-content' into gh/add-pagination-data-models

gherceg committed:
create upstream/downstream urls in django, not frontend

gherceg committed:
don't capitalize Url

gherceg committed:
Merge pull request #29883 from dimagi/gh/update-text-on-release-content

[ERM UI] Update text on downstream linked projects view

gherceg committed:
Merge pull request #29891 from dimagi/gh/add-pagination-data-models

[ERM UI] Add search and pagination to pull release content tab

gherceg committed:
manually remove bootstrap modal elements

gherceg committed:
Merge pull request #29925 from dimagi/gh/fix-b3-modal-bug

Fix remove link modal bug

gherceg committed:
address feedback

gherceg committed:
replace release with push

gherceg committed:
Merge pull request #29932 from dimagi/gh/update-release-content-page

[ERM UI] Update release content page to match mock ups

gherceg committed:
filter domain_links in js

- create push content view model to house localDomainLinks array
- move other push related data to this view model
- rename release to push where applicable

gherceg committed:
stickler

gherceg committed:
build view model no matter what

gherceg committed:
add urls to hyperlinked text

gherceg committed:
Merge pull request #29952 from dimagi/gh/add-doc-links

Add urls to hyperlinks

gherceg committed:
Merge branch 'gh/improve-erm-ui' into gh/update-push-content-domains

gherceg committed:
Merge pull request #29945 from dimagi/gh/update-push-content-domains

[ERM UI] Update domains to push to if a new domain is added

gherceg committed:
move if check to outer most div

gherceg committed:
Merge pull request #29961 from dimagi/gh/hide-downstream-projects-content

Hide downstream projects content

gherceg committed:
Merge branch 'master' into gh/improve-erm-ui

orangejenny reviewed (approved):
Approving since all branches this is built off of have been reviewed.

gherceg committed:
initial work on landing page

gherceg committed:
implement make downstream project flow

gherceg committed:
refactor to one binding

gherceg committed:
refactor django template ifs to ko ifs

gherceg committed:
use computed observables for cleanliness

gherceg committed:
sort upstream domains

gherceg committed:
add blurb on add downstream project modal

gherceg committed:
isort and remove unused method

gherceg committed:
stickler

gherceg committed:
Merge branch 'master' into gh/improve-erm-ui

gherceg committed:
Merge branch 'gh/improve-erm-ui' into gh/linked-project-landing-page

gherceg committed:
semi-colon

gherceg committed:
Merge branch 'master' into gh/improve-erm-ui

gherceg committed:
Merge branch 'master' into gh/improve-erm-ui

gherceg committed:
accidentally left this in merge conflict resolution

gherceg committed:
Merge branch 'gh/improve-erm-ui' into gh/linked-project-landing-page

gherceg committed:
limit GAed version to enterprise and admin users

gherceg committed:
default object manager for DomainLink already excludes deleted links

gherceg committed:
update is active upstream domain method and clear cache

gherceg committed:
address feedback

- remove unnecessary </br>
- update name for status var

gherceg committed:
Merge pull request #30052 from dimagi/gh/linked-project-landing-page

[ERM UI] Landing page if domain contains no links

gherceg committed:
Merge branch 'gh/improve-erm-ui' into gh/change-query-to-enterprise-domains

gherceg committed:
remove unnecessary computed var

ShivamPokhriyal committed:
Merge branch 'master' of https://github.com/dimagi/commcare-hq into gh/improve-erm-ui

gherceg committed:
address feedback

- include obj in domain vars
- use explicit domain names

gherceg committed:
Merge pull request #30071 from dimagi/gh/remove-unnecessary-computed-var

Remove computed var

gherceg committed:
Merge branch 'gh/change-query-to-enterprise-domains' into gh/base-linked-project-access

gherceg committed:
Merge branch 'master' into gh/improve-erm-ui

gherceg committed:
only show release management tab if domain has privilege and user is admin

gherceg committed:
hide ui elements

- show no domain links page if no privilege enabled
- hide add downstream button if no privilege enabled

gherceg committed:
fix failing test

gherceg committed:
Merge pull request #30086 from dimagi/gh/fix-tests

Fix failing test on base branch

gherceg committed:
add util method and view decorators to limit access

gherceg committed:
Merge branch 'master' into gh/improve-erm-ui

gherceg committed:
Merge branch 'gh/improve-erm-ui' into gh/change-query-to-enterprise-domains

gherceg committed:
replace linked_projects with release_management

gherceg committed:
Merge branch 'gh/improve-erm-ui' into gh/base-linked-project-access

gherceg committed:
refactor tests to check for non-enterprise privilege

NOTE: this will be rolled out to all enterprise accounts in the near future
so will have to change back soon

gherceg committed:
Merge branch 'gh/change-query-to-enterprise-domains' into gh/base-linked-project-access

gherceg committed:
Merge branch 'gh/base-linked-project-access' into gh/show-erm-tab-if-applicable

gherceg committed:
replace linked_projects with release_management

gherceg committed:
Merge branch 'gh/base-linked-project-access' into gh/hide-ui-elements-if-applicable

gherceg committed:
replace linked_projects with release_management

gherceg committed:
Merge pull request #30081 from dimagi/gh/hide-ui-elements-if-applicable

Linked projects UI depends on feature flag vs privilege

gherceg committed:
Merge branch 'master' into gh/improve-erm-ui

gherceg committed:
clear cache for upstream domain, not downstream

gherceg committed:
Merge pull request #30101 from dimagi/gh/do-not-show-deleted-links

Clear cache for upstream domain, not downstream

gherceg committed:
check if push tab content is active

gherceg committed:
move comment

gherceg committed:
Merge branch 'gh/improve-erm-ui' into gh/change-query-to-enterprise-domains

gherceg committed:
Merge pull request #30063 from dimagi/gh/change-query-to-enterprise-domains

Only return domains that are eligible to link downstream

gherceg committed:
Merge branch 'gh/improve-erm-ui' into gh/base-linked-project-access

gherceg committed:
Merge branch 'gh/base-linked-project-access' into gh/show-erm-tab-if-applicable

gherceg committed:
update text

gherceg committed:
update tabclass names

gherceg committed:
missed one

gherceg committed:
update text for enterprise release management

gherceg committed:
remove sticky tabs

gherceg committed:
use unique report dispatcher for release management related features

gherceg committed:
Merge pull request #30105 from dimagi/gh/update-linked-domain-text

Linked Project Space UI Improvements

gherceg committed:
Merge pull request #30103 from dimagi/gh/fix-tab-issue

Fix tab issue

gherceg committed:
Merge branch 'gh/improve-erm-ui' into gh/base-linked-project-access

gherceg committed:
Merge branch 'master' into gh/improve-erm-ui

gherceg committed:
Merge branch 'gh/improve-erm-ui' into gh/base-linked-project-access

gherceg committed:
Merge branch 'gh/base-linked-project-access' into gh/show-erm-tab-if-applicable

gherceg committed:
disable checkbox if necessary

- if RELEASE_MANAGEMENT privilege enabled, only able to link apps with active
downstream domains
- behavior should remain the same for LINKED_DOMAINS toggle users

gherceg committed:
keep same behavior as before for linked_domains users

gherceg committed:
enforce admin for RELEASE_MANAGEMENT version

gherceg committed:
address Biyeun's feedback

gherceg committed:
add tests for underlying method in decorator; fix tests

gherceg committed:
Merge branch 'gh/limit-domains-in-app-manager' into gh/show-erm-tab-if-applicable+gh/limit-domains-in-app-manager

gherceg committed:
add migration for remote_base_url help_text

gherceg committed:
add user :facepalm:

gherceg committed:
Merge branch 'gh/limit-domains-in-app-manager' into gh/show-erm-tab-if-applicable+gh/limit-domains-in-app-manager

gherceg committed:
should have used request.couch_user

gherceg committed:
Merge branch 'gh/limit-domains-in-app-manager' into gh/show-erm-tab-if-applicable+gh/limit-domains-in-app-manager

gherceg committed:
Merge pull request #30124 from dimagi/gh/help-text-migration

Add migration for remote_base_url help_text field

gherceg committed:
check for superuser access

gherceg committed:
break up method to get available domains to link

gherceg committed:
Merge branch 'gh/limit-domains-in-app-manager' into gh/show-erm-tab-if-applicable+gh/limit-domains-in-app-manager

gherceg committed:
update message text

gherceg committed:
Merge branch 'gh/limit-domains-in-app-manager' into gh/show-erm-tab-if-applicable+gh/limit-domains-in-app-manager

gherceg committed:
Merge branch 'master' into gh/improve-erm-ui

gherceg committed:
Merge branch 'gh/improve-erm-ui' into gh/base-linked-project-access

gherceg committed:
Merge branch 'gh/base-linked-project-access' into gh/show-erm-tab-if-applicable

gherceg committed:
Merge branch 'gh/base-linked-project-access' into gh/limit-domains-in-app-manager

gherceg committed:
Merge branch 'gh/show-erm-tab-if-applicable' into gh/show-erm-tab-if-applicable+gh/limit-domains-in-app-manager

gherceg committed:
Merge branch 'gh/limit-domains-in-app-manager' into gh/show-erm-tab-if-applicable+gh/limit-domains-in-app-manager

gherceg committed:
error in tests

gherceg committed:
Merge branch 'gh/limit-domains-in-app-manager' into gh/show-erm-tab-if-applicable+gh/limit-domains-in-app-manager

gherceg committed:
update text on linked project space history report

gherceg committed:
Merge pull request #30130 from dimagi/gh/update-linked-project-history-text

Update text on Linked Project Space History report

gherceg committed:
Merge branch 'master' into gh/improve-erm-ui

gherceg committed:
Merge branch 'gh/improve-erm-ui' into gh/base-linked-project-access

gherceg committed:
Merge branch 'gh/base-linked-project-access' into gh/show-erm-tab-if-applicable

gherceg committed:
Merge branch 'gh/show-erm-tab-if-applicable' into gh/show-erm-tab-if-applicable+gh/limit-domains-in-app-manager

gherceg committed:
is_domain_admin already checks for superuser; remove tests

gherceg committed:
dynamic options for multiselect custom binding

gherceg committed:
address Jenny's feedback

- update documentation for billing_account
- use BillingAccount.get_account_by_domain instead

gherceg committed:
js stickler issues

gherceg committed:
Merge pull request #30116 from dimagi/gh/limit-domains-in-app-manager

Disable Copy as Linked Application when appropriate

gherceg committed:
Merge branch 'gh/limit-domains-in-app-manager' into gh/show-erm-tab-if-applicable+gh/limit-domains-in-app-manager

gherceg committed:
Merge branch 'gh/base-linked-project-access' into gh/show-erm-tab-if-applicable

gherceg committed:
limit upstream domains based on account or user

gherceg committed:
cleanup

gherceg committed:
Merge pull request #30135 from dimagi/gh/multiselectOptions-binding

Dynamic options for multiselect custom binding

gherceg committed:
improve naming

gherceg committed:
Merge pull request #30138 from dimagi/gh/limit-upstream-domains-to-account

Limit upstream domains based on account vs user

gherceg committed:
Merge pull request #30077 from dimagi/gh/show-erm-tab-if-applicable

Add ERM section to Project Settings with appropriate permissions

gherceg committed:
Merge branch 'gh/base-linked-project-access' into gh/improve-erm-ui

gherceg committed:
only show one active tab

gherceg committed:
only limit to linked domains if erm enabled

gherceg committed:
allow superusers to still access any domain

gherceg committed:
Merge pull request #30151 from dimagi/gh/fix-3tab-issue

Fix conflicting active tabs on linked projects page

gherceg committed:
use currentTarget; add new var to initialPageData

gherceg committed:
Merge pull request #30152 from dimagi/gh/only-disable-linked-apps-for-erm

Only disable linked apps for release management privilege

gherceg committed:
fix js issue with remote reports

gherceg committed:
it is a method you dumdum

gherceg committed:
add rmi decorator

gherceg committed:
Merge pull request #30185 from dimagi/gh/permissions-for-erm-section

Fix access control for release management/linked projects

gherceg committed:
Merge pull request #30186 from dimagi/gh/fix-js-issue

Fix remote reports issues

gherceg committed:
Merge branch 'master' into gh/improve-erm-ui

gherceg committed:
bump migration

gherceg committed:
track domain link creation and pushes

gherceg committed:
minor stickler comments

gherceg committed:
Merge pull request #30194 from dimagi/gh/address-stickler

Minor stickler comments in domain_links.js

gherceg committed:
pass data as kwargs

gherceg committed:
is_domain_admin is a method

gherceg committed:
rename pulled data model

gherceg committed:
Merge pull request #30193 from dimagi/gh/track-link-creation

Add domain link create and push events to kissmetrics

gherceg committed:
Merge pull request #30195 from dimagi/gh/fix-mock-test-issue

Fix mock test issue
